clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -417,7 +497,10 @@ public void unmute(String id) {
     }
 
     private void unmute(KafkaChannel channel) {
-        channel.unmute();
+        explicitlyMutedChannels.remove(channel);
+        if (!outOfMemory) { //only if we're not muting everything because no memory
+            channel.unmute();
```

Comment:
Similar to the mute in `poll()` - the mute could be delayed until a buffer needs to be allocated? It is possible that the channel already has a buffer allocated, in which case, we want it to complete read.